src/main/java/picard/sam/util/SamComparison.java

```diff
@@ -1,67 +1,153 @@
+        markDuplicatesCheckLeft = SortingCollection.newInstance(SAMRecord.class, new BAMRecordCodec(leftHeader),
+                leftDupComparator, SAMFileWriterImpl.getDefaultMaxRecordsInRam());
+        markDuplicatesCheckRight = SortingCollection.newInstance(SAMRecord.class, new BAMRecordCodec(rightHeader),
+                rightDupComparator, SAMFileWriterImpl.getDefaultMaxRecordsInRam());
+    }
+
+    private void countLenientDuplicateMarkingDifferences() {
+        /* Count duplicate marking differences allowing swaps within duplicate sets.  Should only be used when lenientDup == true.
+           Reads which have different duplicate markings will have been saved in markDuplicatesCheckLeft and markDuplicatesCheckRight
+           in catalogDuplicateDifferences.
+         */
+
+        if (!samComparisonArgumentCollection.LENIENT_DUP) {
+            throw new PicardException("Should only use countLenientDuplicateMarkingDifferences when in lenient duplicate marking mode.");
+        }
+
+        final DuplicateSetIterator duplicateSetLeftIterator = new DuplicateSetIterator(markDuplicatesCheckLeft.iterator(), leftReader.getFileHeader(), true);
+        final DuplicateSetIterator duplicateSetRightIterator = new DuplicateSetIterator(markDuplicatesCheckRight.iterator(), rightReader.getFileHeader(), true);
+        /* Iterate through duplicateSetIterators.  This will iterate through all reads with differing duplicate markings in left and right files.  Passing through the first
+         * iterator (i.e. left file), for each duplicate set we create a set of duplicate reads that the representative read could "swap" with.  On the pass through the second iterator, for each
+         * duplicate read, we look for the reads it is allowed to swap with in the duplicate set from the first iterator.  If the allowed swap can also be performed in this duplicate set,
+         * we do not count the read or the read it could be swapped with as having mismatched duplicates.  Note however, that once a swap is allowed, the reads swapped cannot be used for any
+         * further swaps.
+         */
+        final HashMap<String, Set<String>> allowedSwapsMap = new HashMap<>();
+        for (final DuplicateSet duplicateSet : new IterableAdapter<>(duplicateSetLeftIterator)) {
+            // do not want to redo duplicate marking, want to keep marks assigned in input files
+            final List<SAMRecord> records = duplicateSet.getRecords(false);
+            if (records.size() > 1) {
+                final Set<String> duplicateRecordNames = records.stream().filter(SAMRecord::getDuplicateReadFlag).map(SAMRecord::getReadName)
+                        .collect(Collectors.toSet());
+                records.stream().filter(r -> !r.getDuplicateReadFlag())
+                        .forEach(r -> allowedSwapsMap.put(r.getReadName(), duplicateRecordNames));
```

Comment:
Isn't it the case that there should only ever be one read in the stream at this point (after the preceding filter) ? If so, perhaps its worth collecting at this point, and then validating or logging a warning if there is more than one read, since otherwise something went wrong earlier ?

Reply:
I think the danger there is that the duplicate marking implemented in `DuplicateSetIterator` might not be identical to the duplicate marking used to create the input files.  For example, if the input files where duplicate marked in a UMI aware manner, a duplicate set found by `DuplicateSetIterator` (which is not UMI aware) may have multiple "representative" reads, since we maintain the duplicate marking flags from the input files, where the single duplicate set could be split into multiple duplicate sets based on UMIs.